File: Utils/Functional.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace Shaftesbury.Functional.Utils
{
    /// <summary>
    /// http://msdn.microsoft.com/en-us/library/ee353738.aspx
    /// </summary>
    public static class Functional
    {
        /// <summary> partition: (A -> bool) -> A list -> A list * A list</summary>
        /// <returns> (list * list). The first list contains all items for which f(a) is true. The second list contains the remainder.</returns>
        public static Tuple<List<A>, List<A>> partition<A>(System.Func<A,bool> pred, IEnumerable<A> input)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (pred == null) throw new ArgumentNullException("pred");
            #endregion
            var left = new List<A>();
            var right = new List<A>();

            foreach (var element in input)
                (pred(element) ? left : right).Add(element);

            return Tuple.Create(left, right);
        }

        public static IEnumerable<IGrouping<K, V>> Partition<K, V>(this IEnumerable<V> input, System.Func<V, K> groupFn)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (groupFn == null) throw new ArgumentNullException("groupFn");
            #endregion
            return from item in input group item by groupFn(item);
        }

        public static IEnumerable<IGrouping<bool, B>> Select<A, B>(this IEnumerable<IGrouping<bool, A>> input, System.Func<A, B> left, System.Func<A, B> right)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (left == null) throw new ArgumentNullException("left");
            if (right == null) throw new ArgumentNullException("right");
            #endregion
            return input.SelectMany(grp => grp, (grp, element) => new { grp.Key, Value = (grp.Key ? left : right)(element) }).GroupBy(item => item.Key, item => item.Value);
        }

        public static IEnumerable<IGrouping<G, B>> Select<A, B, G>(this IEnumerable<IGrouping<G, A>> input, IDictionary<G, System.Func<A, B>> transformations)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (transformations == null) throw new ArgumentNullException("transformations");
            #endregion
            return input.Where(item => transformations.ContainsKey(item.Key)).
                SelectMany(grp => grp, (grp, element) => new { grp.Key, Value = transformations[grp.Key](element) }).
                GroupBy(item => item.Key, item => item.Value);
        }

        // An interesting alternative to using a HashSet
        public static IEnumerable<A> Unique<A,Key>(this IEnumerable<A> input, Func<A,Key> f)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return input.GroupBy<A, Key>(f).SelectMany(i => i.Take(1));
        }

        public static B Try<A, B, E>(this A input, System.Func<A, B> tryClause, System.Func<A, B> catchClause) where E:Exception
        {
            #region Precondition
            if (tryClause == null) throw new ArgumentNullException("tryClause");
            if (catchClause == null) throw new ArgumentNullException("catchClause");
            #endregion
            B results;
            try
            {
                results = tryClause(input);
            }
            catch (E)
            {
                results = catchClause(input);
            }
            return results;
        }

        public static B If<A, B>(this A a, System.Func<A,bool> predicate, System.Func<A, B> thenClause, System.Func<A, B> elseClause)
        {
            #region Precondition
            if (a == null) throw new ArgumentNullException("a");
            if (predicate == null) throw new ArgumentNullException("predicate");
            if (thenClause == null) throw new ArgumentNullException("thenClause");
            if (elseClause == null) throw new ArgumentNullException("elseClause");
            #endregion
            return predicate(a) ? thenClause(a) : elseClause(a);
        }

        public class Case<A, B>
        {
            public System.Func<A,bool> check { get; internal set; }
            public System.Func<A, B> results { get; internal set; }
        }

        public static B Switch<A, B>(this A input, IEnumerable<Case<A, B>> cases, System.Func<A, B> defaultCase)
        {
            #region Precondition
            if (cases == null) throw new ArgumentNullException("cases");
            if (defaultCase == null) throw new ArgumentNullException("defaultCase");
            #endregion
            return Try<InvalidOperationException>.ToTry(input, a => cases.First(chk => chk.check(a)).results(a), defaultCase);
        }

	    public static IEnumerable<B> Choose<A,B>(this IEnumerable<A> input, System.Func<A,Maybe<B>> chooseFn)
	    {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (chooseFn == null) throw new ArgumentNullException("chooseFn");
            #endregion
            return input.Select(chooseFn).OfType<Something<B>>().Select(intermediate => intermediate.Value);
	    }

        public static IEnumerable<B> Choose<A, B>(this IEnumerable<A> input, System.Func<A, Option<B>> chooseFn)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (chooseFn == null) throw new ArgumentNullException("chooseFn");
            #endregion
            return input.Select(chooseFn).Where(o => o.IsSome).Select(intermediate => intermediate.Some);
        }

        /// <summary> sortWith: (A -> A -> int) -> A list -> A list</summary>
        public static List<A> sortWith<A>(Comparison<A> f, List<A> input)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            List<A> output = new List<A>(input);    
            output.Sort(f);
            return output;
        }

        /// <summary> not: (A -> bool) -> (A -> bool)</summary>
        public static System.Func<A, bool> not<A>(System.Func<A, bool> f)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return a => !f(a);
        }

        /// <summary> not2: (A -> B -> bool) -> (A -> B -> bool)</summary>
        public static System.Func<A,B,bool> not2<A, B>(System.Func<A,B,bool> f)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return (a, b) => !f(a, b);
        }


        /// <summary> forAll2: (A -> B -> bool) -> A list -> B list -> bool</summary>
        public static bool forAll2<A, B>(System.Func<A,B,bool> f, IEnumerable<A> input1, IEnumerable<B> input2)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            if (input1 == null) throw new ArgumentNullException("input1");
            if (input2 == null) throw new ArgumentNullException("input2");
            #endregion
            IEnumerator<A> enum1 = input1.GetEnumerator();
            IEnumerator<B> enum2 = input2.GetEnumerator();
            bool enum1Moved = false, enum2Moved = false;
            do
            {
                enum1Moved = enum1.MoveNext();
                enum2Moved = enum2.MoveNext();
                if (enum1Moved && enum2Moved && !f(enum1.Current, enum2.Current))
                    return false;
            } while (enum1Moved && enum2Moved);
            if( enum1Moved != enum2Moved)
                throw new ArgumentException();
            return true;
        }

        /// <summary> pick: (A -> B option) -> A list -> B</summary>
        public static B pick<A, B>(System.Func<A, Maybe<B>> f, IEnumerable<A> input)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            if (input == null) throw new ArgumentNullException("input");
            #endregion
            return input.Choose(f).First();
        }

        public static Tuple<List<A>,List<B>> Unzip<A,B>(this IEnumerable<Tuple<A,B>> input)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            #endregion
            var left = new List<A>();
            var right = new List<B>();
            foreach(var tuple in input)
            {
                left.Add(tuple.Item1);
                right.Add(tuple.Item2);
            }
            return Tuple.Create(left, right);
        }

        /// <summary> Zip3: A list -> B list -> C list -> (A * B* C) list</summary>
        public static IEnumerable<Tuple<A,B,C>> Zip3<A, B, C>(IEnumerable<A> input1, IEnumerable<B> input2, IEnumerable<C> input3)
        {
            #region Precondition
            if (input1 == null) throw new ArgumentNullException("input1");
            if (input2 == null) throw new ArgumentNullException("input2");
            if (input3 == null) throw new ArgumentNullException("input3");
            #endregion
            IEnumerator<A> enum1 = input1.GetEnumerator();
            IEnumerator<B> enum2 = input2.GetEnumerator();
            IEnumerator<C> enum3 = input3.GetEnumerator();
            bool enum1Moved, enum2Moved, enum3Moved;
            do
            {
                enum1Moved = enum1.MoveNext();
                enum2Moved = enum2.MoveNext();
                enum3Moved = enum3.MoveNext();
                if (enum1Moved && enum2Moved && enum3Moved)
                    yield return Tuple.Create(enum1.Current, enum2.Current, enum3.Current);
            } while (enum1Moved && enum2Moved & enum3Moved);
            if (enum1Moved != enum2Moved || enum1Moved != enum3Moved || enum2Moved != enum3Moved)
                throw new ArgumentException();
        }

        public static IEnumerable<A> Unique<A>(this IEnumerable<A> input)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            #endregion
            return input.GroupBy(i => i).SelectMany(i => i.Take(1));
        }


        #region Standard predicates
        public static bool IsOdd(int v) { return v % 2 != 0; }
        public static bool IsEven(int v) { return v % 2 == 0; }
        public static bool IsNull<T>(T t) where T:class { return t == null; }
        public static bool IsNotNull<T>(T t) where T : class { return t != null; }
        
        public static bool AreEqual<A, B>(A a, B b) { return a.Equals(b); }

        public static System.Func<int,bool> dIsOdd = IsOdd;
        public static System.Func<int,bool> dIsEven = IsEven;
        public static System.Func<object,bool> dIsNull = IsNull;
        public static System.Func<object,bool> dIsNotNull = IsNotNull;
        #endregion

        #region Standard initialisers
        public static System.Func<int,T> Constant<T>(T i) { return j => i; }
        #endregion

        #region Standard map functions
        public static string Stringify(int a) { return a.ToString(); }
        public static System.Func<string, string> dQuote = s => "'" + s + "'";
        public static System.Func<int, string> dStringify = Stringify;
        #endregion

        #region Standard fold functions
        private static int Sum(int state, int b) { return state + b; }
        private static int Max(int state, int b) { return state < b ? b : state; }
        private static int Min(int state, int b) { return state > b ? b : state; }
        private static int Count(int state, int b) { return state + 1; }

        public static System.Func<int, int, int> dSum = Sum;
        public static System.Func<int, int, int> dMax = Max;
        public static System.Func<int, int, int> dMin = Min;
        public static System.Func<int, int, int> dCount = Count;
        #endregion

        #region Standard sort functions
        public static int Sorter<A>(A left, A right) where A:IComparable
        {
            return left.CompareTo(right);
        }
        #endregion

        /// <summary>Use Func for forward pipelining of lists and curried functions</summary>
        /// <typeparam name="A"></typeparam>
        /// <typeparam name="RetType"></typeparam>
        //public class Func<A, RetType>
        //{
        //    public static RetType operator >(A a, Func<A, RetType> f)
        //    {
        //        return f.act(a);
        //    }
        //    /// <summary>Not implemented because it makes no sense. C# compiler requires the matching operator but logic does not</summary>
        //    public static RetType operator <(A a, Func<A, RetType> f)
        //    {
        //        throw new NotImplementedException();
        //    }

        //    public RetType act(A a)
        //    {
        //        return _f(a);
        //    }

        //    private readonly System.Func<A, RetType> _f;
        //    public Func(System.Func<A, RetType> f)
        //    {
        //        _f = f;
        //    }
            
        //    public static implicit operator System.Func<A,RetType> (Func<A, RetType> f)
        //    {
        //        return f._f;
        //    }
        //    public static implicit operator Func<A, RetType> (System.Func<A, RetType> f)
        //    {
        //        return new Func<A, RetType>(f);
        //    }
        //}

        /* 
        // Curried functions for use with Func<A,B> 
        */
        public static Func<IEnumerable<A>, IEnumerable<A>> filter<A>(System.Func<A,bool> f)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return l => l.Where(f);
        }

        public static Func<IEnumerable<A>, IEnumerable<B>> map<A, B>(System.Func<A,B> f)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return l => l.Select(f);
        }

        /// <summary> choose: (A -> B option) -> A list -> B list</summary>
        public static Func<IEnumerable<A>, IEnumerable<B>> choose<A, B>(System.Func<A, Maybe<B>> f) where B : class
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return l => l.Choose(f);
        }

        /// <summary> init: int -> (int -> A) -> A list</summary>
        public static IEnumerable<T> Repeat<T>(this System.Func<int, T> f, int howMany)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return Enumerable.Range(0, howMany).Select(f);
        }

        /// <summary> init: int -> (int -> A) -> A list</summary>
        // ReSharper disable FunctionNeverReturns
        public static IEnumerable<T> Repeat<T>(this System.Func<int,T> f)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            for (int i = 0; ; ++i)
                yield return f(i);
        }
        // ReSharper restore FunctionNeverReturns
    
        public class Enumerate<A>:IEnumerable<A>
        {
            private readonly IEnumerator<A> enumerator;
            public Enumerate(IEnumerator<A> en)
            {
                enumerator = en;
            }
            #region Implementation of IEnumerable

            public IEnumerator<A> GetEnumerator()
            {
                return enumerator;
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                return GetEnumerator();
            }

            #endregion
        }

        public static Tuple<List<A>,IEnumerable<A>> TakeNAndYield<A>(this IEnumerable<A> input, int howMany)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            #endregion
            int counter = 0;
            var output = new List<A>();
            var enumerator = input.GetEnumerator();
            if(howMany>0&&enumerator.MoveNext())
            {
                while(counter<howMany)
                {
                    output.Add(enumerator.Current);
                    counter++;
                    if (counter < howMany && !enumerator.MoveNext()) break;
                }
                return Tuple.Create(output, new Enumerate<A>(enumerator).AsEnumerable());
            }
            return Tuple.Create(output, new A[] {}.AsEnumerable());
        }
    }

    public static class Case
    {
        public static Functional.Case<A, B> ToCase<A, B>(System.Func<A,bool> pred, Func<A, B> res)
        {
            #region Precondition
            if (pred == null) throw new ArgumentNullException("pred");
            if (res == null) throw new ArgumentNullException("res");
            #endregion
            return new Functional.Case<A, B> { check = pred, results = res };
        }
    }

    public static class Try<E> where E:Exception
    {
        public static B ToTry<A,B>(A input, Func<A, B> tryClause, Func<A, B> catchClause)
        {
            #region Precondition
            if (tryClause == null) throw new ArgumentNullException("tryClause");
            if (catchClause == null) throw new ArgumentNullException("catchClause");
            #endregion
            return Functional.Try<A, B, E>(input, tryClause, catchClause);
        }
    }

    public static class ToIgnore
    {
        public static void Ignore<T>(this T t)
        {
            var input = t as IEnumerable;
            if(input!=null)
                foreach (var i in input)
                    i.Ignore();
        }
    }

    public static class ToFunc
    {
        public static B In<A, B>(this A a, Func<A, B> f)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            #endregion
            return f(a);
        }

        public static Func<A, C> Then<A, B, C>(this Func<A, B> f, Func<B, C> g)
        {
            #region Precondition
            if (f == null) throw new ArgumentNullException("f");
            if (g == null) throw new ArgumentNullException("g");
            #endregion
            return a => g(f(a));
        }

        /// <summary>
        /// Similarly to In and Then, use this in a chain when building objects, eg
        /// var a = Bob.Build(Today).ThenTransformIfTrue(MakeItBigger, i => true)
        /// </summary>
        public static T ThenTransformIfTrue<T, U>(this T input, Func<T, U> tfm, Func<T, bool> pred) where U : T
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            if (tfm == null) throw new ArgumentNullException("tfm");
            if (pred == null) throw new ArgumentNullException("pred");
            #endregion
            return pred(input) ? tfm(input) : input;
        }
    }

    public static class Converter
    {
        public static HashSet<T> ToHashSet<T>(this IEnumerable<T> input)
        {
            #region Precondition
            if (input == null) throw new ArgumentNullException("input");
            #endregion
            return new HashSet<T>(input);
        }
    }

    public static class Numerical
    {
        public static bool Between<A>(this A item, A left, A right) where A : IComparable<A>
        {
            #region Precondition
            if (item == null) throw new ArgumentNullException("item");
            #endregion
            return item.CompareTo(left) >= 0 && item.CompareTo(right) <= 0;
        }
    }
}

File: Functional-Utils-Java/src/me/shaftesbury/utils/functional.java
/**
 * Created with IntelliJ IDEA.
 * User: Bob
 * Date: 16/10/13
 * Time: 20:04
 * To change this template use File | Settings | File Templates.
 */
package me.shaftesbury.utils;

import org.javatuples.Pair;

import java.util.*;

public final class functional
{
    private functional() {}

    public static boolean isNullOrEmpty(final String s)
    {
        return s==null || s.isEmpty();
    }

    public static <T>String join(final String delimiter, final Iterable<T> strs)
    {
        if(strs==null) return "";
        Iterator<T> it = strs.iterator();
        StringBuilder sb = new StringBuilder();
        boolean isFirst = true;
        while(it.hasNext())
        {
            if(!isFirst) sb.append(delimiter);
            sb.append(it.next());
            isFirst=false;
        }
        return sb.toString();
    }

    public static String indentBy(final int howMany, final String unitOfIndentation, final String indentThis)
    {
        Collection<String> indentation = init(
                new Func<Integer, String>() {
                    @Override
                    public String apply(Integer integer) {
                        return unitOfIndentation;
                    }
                }, howMany);
        return fold(new Func2<String, String, String>() {
            @Override
            public String apply(String state, String str) {
                return str + state;
            }
        },indentThis, indentation);
    }

    public static <A, B>Pair<A,Collection<B>> foldAndChoose(
        Func2<A, B, Pair<A,Option<B>>> f,
        A initialValue, Iterable<B> input) throws Exception
    {
        if (f == null) throw new /*ArgumentNull*/Exception("f");
        if (input == null) throw new /*ArgumentNull*/Exception("input");

        A state = initialValue;
        Collection<B> results = new ArrayList<B>();
        for (B b : input)
        {
            Pair<A, Option<B>> intermediate = f.apply(state, b);
            state = intermediate.getValue0();
            if (!intermediate.getValue1().isNone())
                results.add(intermediate.getValue1().Some());
        }
        return new Pair<A, Collection<B>>(state, results);
    }

    public static final <T>List<T> convert(final Enumeration<T> input)
    {
        final List<T> output = new ArrayList<T>();
        while(input.hasMoreElements())
            output.add(input.nextElement());
        return output;
    }

    /// <summary>
    /// Analogue of string.Join for List&lt;T&gt; with the addition of a user-defined map function
    /// </summary>
    /// <typeparam name="T"></typeparam>
    /// <param name="separator"></param>
    /// <param name="l"></param>
    /// <param name="fn"></param>
    /// <returns></returns>
    public static <T>String join(String separator, Iterable<T> l, Func<T, String> fn) throws Exception
    {
        if (l == null) throw new /*ArgumentNull*/Exception("l");
        if (fn == null) throw new /*ArgumentNull*/Exception("fn");

        return join(separator, map(fn,l));
    }

    /// <summary>return lowerBound &lt; val &lt; upperBound</summary>
    public static <T extends Comparable<T>>boolean between(T lowerBound, T upperBound, T val) throws Exception
    {
        if (val == null) throw new /*ArgumentNull*/Exception("val");

        return val.compareTo(lowerBound) == 1 && val.compareTo(upperBound) == -1;
    }

    /// <summary> findLast: (A -> bool) -> A list -> A</summary>
    public static <A>A findLast(Func<A,Boolean> f, List<A> input) throws Exception
    {
        if (f == null) throw new /*ArgumentNull*/Exception("f");
        if (input == null) throw new /*ArgumentNull*/Exception("input");

        for (A a : Enumerators.ReverseEnum(input))
            if (f.apply(a))
                return a;
        throw new /*KeyNotFound*/Exception();
    }

    public interface Func<A,R>
    {
        public R apply(final A a);
    }

    public static <A, B> B In( final A input, final Func<A, B> f)
    {
        return f.apply(input);
    }

    public static <A, B, C> Func<A, C> Then(final Func<A, B> f, final Func<B, C> g)
    {
        return new Func<A, C>()
        {
            @Override
            public C apply(A x)
            {
                return g.apply(f.apply(x));
            }
        };
    }

    public static final Func<Integer,Boolean> IsEven = new Func<Integer, Boolean>()
    {
        @Override
        public Boolean apply(Integer i)
        {
            return i % 2 == 0;
        }
    };

    public static final Func<Integer,Boolean> IsOdd = new Func<Integer, Boolean>()
    {
        @Override
        public Boolean apply(Integer i)
        {
            return i % 2 != 0;
        }
    };

    public interface Func2<A,B,C> // disappointing! This should be related to Func<A,Func<B,C>>
    {
        public C apply(final A a, final B b);
    }


    /// <summary> init: int -> (int -> A) -> A list</summary>
    public static <T>Collection<T> init(final Func<Integer,T> f,final int howMany)
    {
        //if (f == null) throw new ArgumentNullException("f");

        Collection<T> output = new ArrayList<T>();
        for(int i=0; i<howMany; ++i)
            output.add(f.apply(i));
        return output;
    }

    /// <summary> map: (A -> B) -> A list -> B list</summary>
    public static <A,B> Collection<B> map(final Func<A, B> f, final Iterable<A> input)
    {
        Collection<B> output = new ArrayList<B>();
        for(A a : input)
            output.add(f.apply(a));
        return output;
    }

    /// <summary> sortWith: (A -> A -> int) -> A list -> A list</summary>
    public static <A>Collection<A> sortWith(final Comparator<A> f, final List<A> input)
    {
        List<A> output = new ArrayList<A>(input);
        Collections.sort(output, f);
        return output;
    }

    public static <A extends Comparable<A>>int Sorter(final A left, final A right)
    {
        return left.compareTo(right);
    }
    public static Comparator<Integer> dSorter = new Comparator<Integer>()
    {
        @Override public int compare(Integer i, Integer j) { return Sorter(i, j); }
    };

    public static <T> String Stringify(final T a) { return a.toString(); }
    public static Func<Integer, String> dStringify = new Func<Integer, String>()
    {
        @Override public String apply(Integer i) { return Stringify(i); }
    };

    /// <summary> forAll2: (A -> B -> bool) -> A list -> B list -> bool</summary>
    public static <A, B>boolean forAll2(final Func2<A, B,Boolean> f, final Iterable<A> input1, final Iterable<B> input2) throws Exception
    {
        Iterator<A> enum1 = input1.iterator();
        Iterator<B> enum2 = input2.iterator();
        boolean enum1Moved = false, enum2Moved = false;
        do
        {
            enum1Moved = enum1.hasNext();
            enum2Moved = enum2.hasNext();
            if (enum1Moved && enum2Moved && !f.apply(enum1.next(), enum2.next()))
                return false;
        } while (enum1Moved && enum2Moved);
        if( enum1Moved != enum2Moved)
            throw new /*Argument*/Exception();
        return true;
    }

    public static <A>Collection<A> filter(final Func<A,Boolean> pred, final Iterable<A> input)
    {
        Collection<A> output = new ArrayList<A>();
        for(A element : input)
        {
            if(pred.apply(element))
                output.add(element);
        }
        return output;
    }

    /// <summary> exists: (A -> bool) -> A list -> bool</summary>
    public static <A>boolean exists(final Func<A,Boolean> f, final Iterable<A> input)
    {
        for(A a : input)
            if(f.apply(a))
                return true;
        return false;
    }

    /// <summary> not: (A -> bool) -> (A -> bool)</summary>
    public static <A>Func<A,Boolean> not(final Func<A,Boolean> f)
    {
        return new Func<A,Boolean>(){@Override public Boolean apply(A a) { return !f.apply(a);}};
    }

    /// <summary> forAll: (A -> bool) -> A list -> bool</summary>
    public static <A>boolean forAll(final Func<A,Boolean> f, final Iterable<A> input)
    {
        return !exists(not(f), input);
    }

    /// <summary> not2: (A -> B -> bool) -> (A -> B -> bool)</summary>
    public static <A,B>Func2<A,B,Boolean> not2(final Func2<A,B,Boolean> f)
    {
        return new Func2<A,B,Boolean>(){@Override public Boolean apply(A a, B b) { return !f.apply(a,b);}};
    }

    /// <summary> partition: (A -> bool) -> A list -> A list * A list</summary>
    /// <returns> (list * list). The first list contains all items for which f(a) is true. The second list contains the remainder.</returns>
    public static <A>org.javatuples.Pair<Collection<A>,Collection<A>> partition(final Func<A,Boolean> f, final Iterable<A> input)
    {
        Collection<A> left = new ArrayList<A>();
        Collection<A> right = new ArrayList<A>();
        for (A a : input)
            if (f.apply(a))
                left.add(a);
            else
                right.add(a);
        return new org.javatuples.Pair<Collection<A>,Collection<A>>(left, right);
    }
    /*
    /// <summary> choose: (A -> B option) -> A list -> B list</summary>
    public static <A, B>Collection<B> choose(Func<A, Option<B>> f, Iterable<A> input)
    {
        Collection<B> results = new ArrayList<B>();
        for(A a : input)
        {
            Option<B> intermediate = f.apply(a);
            if (!intermediate.None)
                results.add(intermediate.Some);
        }
        return results;
    }
    */

    /// <summary> fold: (A -> B -> A) -> A -> B list -> A</summary>
    public static <A, B>A fold(Func2<A, B, A> f, A initialValue, Iterable<B> input)
    {
        A state = initialValue;
        for (B a : input)
            state = f.apply(state, a);
        return state;
    }
}